azure/services/aso/aso.go

```diff
@@ -0,0 +1,165 @@
+)
+
+// Service is an implementation of the Reconciler interface. It handles creation
+// and deletion of resources using ASO.
+type Service struct {
+	client.Client
+}
+
+// New creates a new ASO service.
+func New(ctrlClient client.Client) *Service {
+	return &Service{
+		Client: ctrlClient,
+	}
+}
+
+// CreateOrUpdateResource implements the logic for creating a new or updating an
+// existing resource with ASO.
+func (s *Service) CreateOrUpdateResource(ctx context.Context, spec azure.ASOResourceSpecGetter, serviceName string) (result genruntime.MetaObject, err error) {
+	ctx, log, done := tele.StartSpanWithLogger(ctx, "aso.Service.CreateOrUpdateResource")
+	defer done()
+
+	resource := spec.ResourceRef()
+	resourceName := resource.GetName()
+	resourceNamespace := resource.GetNamespace()
+
+	log = log.WithValues("service", serviceName, "resource", resourceName, "namespace", resourceNamespace)
+
+	var existing genruntime.MetaObject
+	if err := s.Client.Get(ctx, client.ObjectKeyFromObject(resource), resource); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return nil, errors.Wrapf(err, "failed to get existing resource %s/%s (service: %s)", resourceNamespace, resourceName, serviceName)
+		}
+	} else {
+		existing = resource
+		log.V(2).Info("successfully got existing resource")
+
+		// Check if there is an ongoing long running operation.
+		conds := existing.GetConditions()
+		i, readyExists := conds.FindIndexByType(conditions.ConditionTypeReady)
+		if !readyExists {
+			return nil, azure.WithTransientError(errors.New("ready status unknown"), 20*time.Second)
+		}
+		var readyErr error
+		if cond := conds[i]; cond.Status != metav1.ConditionTrue {
+			switch {
+			case cond.Reason == conditions.ReasonReconciling.Name:
+				readyErr = newOperationNotDoneError(existing)
+			default:
+				readyErr = fmt.Errorf("resource is not Ready: %s", conds[i].Message)
+			}
+
+			if readyErr != nil {
+				if conds[i].Severity == conditions.ConditionSeverityError {
+					return nil, azure.WithTerminalError(readyErr)
+				}
+				return nil, azure.WithTransientError(readyErr, 20*time.Second)
+			}
+		}
+	}
+
+	// Construct parameters using the resource spec and information from the existing resource, if there is one.
+	parameters, err := spec.Parameters(ctx, deepCopyOrNil(existing))
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get desired parameters for resource %s/%s (service: %s)", resourceNamespace, resourceName, serviceName)
+	}
+
+	parameters.SetName(resourceName)
+	parameters.SetNamespace(resourceNamespace)
+
+	diff := cmp.Diff(existing, parameters)
+	if diff == "" {
+		log.V(2).Info("resource up to date")
+		return existing, nil
+	}
+
+	// Create or update the resource with the desired parameters.
+	logMessageVerbPrefix := "creat"
+	if existing != nil {
+		logMessageVerbPrefix = "updat"
+	}
+	log.V(2).Info(logMessageVerbPrefix+"ing resource", "diff", diff)
+	if existing != nil {
+		err = s.Client.Update(ctx, parameters)
```

Comment:
Since we're patching a k8s object here, have you looked into using CAPI's patch helper instead of directly calling Create/Update?

Reply:
It looks like that works for update, but we'll still need to call create. I'll add that here.